src/CameraEnvironmental.h

```diff
@@ -3,5 +3,46 @@
 #include "ICamera.h"
 #include "ray.h"
 
-// --- IMPLENET class CCameraEnvironmental ---
+#define _USE_MATH_DEFINES
+#include <math.h>
+// --- IMPLEMENT class CCameraEnvironmental ---
 // --- PUT YOUR CODE HERE ---
+
+class CCameraEnvironmental : public ICamera
+{
+public:
+	CCameraEnvironmental(Size resolution, const Vec3f& pos, float angle)
+		: ICamera(resolution)
+		, m_pos(pos)
+	{
+	}
+	virtual ~CCameraEnvironmental(void) = default;
+
+	virtual void InitRay(Ray& ray, int x, int y) override
+	{
+		// --- PUT YOUR CODE HERE ---
+		float dx = x;
+		float dy = y;
+
+		auto height = getResolution().height;
+		auto width = getResolution().width;
+
+		double theta = 2 * M_PI * dx / height;
+		double phi = acos(1 - 2 * dy / width);
+		double x1 = sin(phi) * cos(theta);
+		double y1 = sin(phi) * sin(theta);
+		double z = cos(phi);
+		Vec3f spherePoint = Vec3f(x1, y1, z);
```

Comment:
this point is on the unit sphere, which is located in the origin. So you can use it as direction vector. Next, you forgot the coordinate system transformation, in the WCS y axis looks up, in the sphere you generate, z axis looks up